File: Contests/CodeforcesPython/Codeforces Round 605 Div. 3/E_Nearest_Opposite_Parity/E_Nearest_Opposite_Parity.go
package main

import (
	"bufio"
	. "fmt"
	"io"
	"os"
	// "sort"
)

type pair struct {
	d, i, t int
}

func CF1272E(_r io.Reader, _w io.Writer) {
	in := bufio.NewReader(_r)
	out := bufio.NewWriter(_w)
	defer out.Flush()

	var n int
	Fscan(in, &n)
	g := make([][]int, n)
	a := make([]int, n)
	ans := make([]int, n)
	q := []pair{}
	for i := range a {
		Fscan(in, &a[i])
		if i+a[i] < n {
			g[i+a[i]] = append(g[i+a[i]], i)
		}
		if i-a[i] >= 0 {
			g[i-a[i]] = append(g[i-a[i]], i)
		}
		q = append(q, pair{0, i, a[i] & 1})
		ans[i] = -1
	}
	visit := make([][]int, 2)
	visit[0] = make([]int, n)
	visit[1] = make([]int, n)

	for len(q) > 0 {
		d, u, t := q[0].d, q[0].i, q[0].t
		q = q[1:]
		for _, v := range g[u] {
			if a[v]&1 != t && (ans[v] == -1 || ans[v] > d+1) {
				ans[v] = d + 1
			}
			if visit[t][v] == 0 || visit[t][v] > d+1 {
				visit[t][v] = d + 1
				q = append(q, pair{d + 1, v, t})
			}
		}
	}

	for _, v := range ans {
		Fprintln(out, v, " ")
	}
}

func main() { CF1272E(os.Stdin, os.Stdout) }
func max(a, b int) int {
	if b > a {
		return b
	}
	return a
}

File: Contests/CodeforcesPython/Codeforces Round 605 Div. 3/E_Nearest_Opposite_Parity/E_Nearest_Opposite_Parity.py
import collections
import math
import os
import random
import sys
from bisect import bisect, bisect_left
from functools import reduce
from heapq import heapify, heappop, heappush
from io import BytesIO, IOBase

# region fastio
BUFSIZE = 8192
class FastIO(IOBase):
    newlines = 0

    def __init__(self, file):
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None

    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()

    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()

    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)

class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")
        
sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
input = lambda: sys.stdin.readline().rstrip("\r\n")
ints = lambda: list(map(int, input().split()))
# endregion fastio

# region interactive
def printQry(a, b) -> None:
    sa = str(a)
    sb = str(b)
    print(f"? {sa} {sb}", flush = True)

def printAns(ans) -> None:
    s = str(ans)
    print(f"! {s}", flush = True)
# endregion interactive

# MOD = 998244353
# MOD = 10 ** 9 + 7
# DIR = ((-1, 0), (0, 1), (1, 0), (0, -1))

def solve() -> None:
    n = int(input())
    arr = ints()
    
    g = [[] for _ in range(n)]
    q = collections.deque()
    for i, a in enumerate(arr):
        if i + a < n: g[i + a].append(i)
        if i - a >= 0: g[i - a].append(i)
        q.append((0, i, a&1))

    visited = [[-1] * 2 for _ in range(n)]

    ans = [-1] * n

    while q:
        step, u, type = q.popleft()
        for v in g[u]:
            if arr[v] & 1 != type and (ans[v] == -1 or ans[v] > step + 1):
                ans[v] = step + 1
            if visited[v][type] == -1 or visited[v][type] > step + 1:
                visited[v][type] = step + 1
                q.append((step + 1, v, type))
    
    print(*ans)


# t = int(input())
# for _ in range(t):
solve()